modules/smartxBidAdapter.js

```diff
@@ -0,0 +1,400 @@
+                adText: 'SmartX Outstream Video Ad via Prebid.js',
+                player_width: playersize[0][0],
+                player_height: playersize[0][1],
+                content_page_url: utils.deepAccess(bidderRequest, 'data.site.page'),
+                ad_mute: +!!utils.deepAccess(currentBidRequest, 'params.ad_mute'),
+                hide_skin: +!!utils.deepAccess(currentBidRequest, 'params.hide_skin'),
+                outstream_options: utils.deepAccess(currentBidRequest, 'params.outstream_options'),
+                outstream_function: utils.deepAccess(currentBidRequest, 'params.outstream_function')
+              }
+            });
+            try {
+              renderer.setRender(outstreamRender);
+              renderer.setEventHandlers({
+                impression: function impression() {
+                  return utils.logMessage('SmartX outstream video impression event');
+                },
+                loaded: function loaded() {
+                  return utils.logMessage('SmartX outstream video loaded event');
+                },
+                ended: function ended() {
+                  utils.logMessage('SmartX outstream renderer video event');
+                }
+              });
+            } catch (err) {
+              utils.logWarn('Prebid Error calling setRender or setEventHandlers on renderer', err);
+            }
+            bid.renderer = renderer;
+          }
+          bidResponses.push(bid);
+        })
+      });
+    }
+    return bidResponses;
+  }
+}
+
+function createOutstreamScript(bid) {
+  // const slot = utils.getBidIdParameter('slot', bid.renderer.config.outstream_options);
+  utils.logMessage('[SMARTX][renderer] Handle SmartX outstream renderer');
+  const elementId = bid.adUnitCode;
+  // eslint-disable-next-line camelcase
+  var sc_smartIntxtStart;
+  // eslint-disable-next-line camelcase
+  var sc_smartIntxtNoad;
+  // eslint-disable-next-line camelcase
+  var sc_smartIntxtEnd;
+  var SmartPlay;
+  let smartPlayObj = {
+    minAdWidth: 290,
+    maxAdWidth: 900,
+    elementLocator: {
+      allowInViewport: false,
+      minimumElementWidth: 290,
+      scanPixelsBelowViewport: 800
+    },
+    onStartCallback: function (m, n) {
+      try {
+        sc_smartIntxtStart(n);
+      } catch (f) {}
+    },
+    onCappedCallback: function (m, n) {
+      try {
+        sc_smartIntxtNoad(n);
+      } catch (f) {}
+    },
+    onEndCallback: function (m, n) {
+      try {
+        sc_smartIntxtEnd(n);
+      } catch (f) {}
+    },
+    debug: true
+  };
+  smartPlayObj.adResponse = bid.vastContent;
+  const script = window.document.createElement('script');
+  script.type = 'text/javascript';
+  script.async = 'true';
+  script.src = 'https://dco.smartclip.net/?plc=7777777';
+  script.onload = script.onreadystatechange = function () {
+    var rs = this.readyState;
+    if (rs && rs != 'complete' && rs != 'loaded') return;
+    try {
+      SmartPlay(elementId, smartPlayObj);
```

Comment:
Hi @jsnellbaker,
the adapter is released, thanks a lot. 

We are facing now problems due to minified code that gets delivered when downloading the adapter from the official page. At this line we create our Outstream-Script Player and we are calling several functions `(sc_smartIntxtStart(n), sc_smartIntxtNoad(n), sc_smartIntxtEnd(n), SmartPlay(elementId, smartPlayObj))` related to our out-stream player but in the minified version they get replaced with `(void 0)(r)` and console error `Prebid ERROR: error caught : TypeError: (void 0) is not a function` so the outstream part doesn't work, what can we do so that the `function createOutstreamScript(bid)` is not changed in the later download version?

Reply:
What if you try to call the script function from the `window` instead of calling it on its own?  Since it doesn't seem the `SmartPlay` variable gets defined within the scope of the bidAdapter code (which is likely why it converted it to the `void(0)` to check if it was undefined), calling from the window sould give it a definite reliable source even in the minified state.

Reply:
We are loading our out-stream playerlibary with the "https://dco.smartclip.net/?plc=7777777" call, there the mentioned functions are available. I have implemented the prebid.js on a test page and replaced the voids with the originally defined functions, everything works fine there.So in our opinion it is only necessary to prevent that the 4 function calls are not changed. Is this possible?

Reply:
The issue with the minified build is that it is likely seeing these variables never get assigned an actual value.  While this isn't the case in a real page (because you expect the script to execute and overwrite the placeholders), it doesn't change from what build stack sees in the code.

In regards to the 3 'callback' functions, why are they defined here in the adapter file and not included as part of the script code?  These functions don't read any values from inside the bidAdapter file to pass along into the script file.  It's just my guess, but the callback functions appear to be internal functions that are within the script file.  Could these 3 functions (`onStartCallback`, `onCappedCallback`, `onEndCallback`) just be taken out of the config object and moved inside the script itself?

For the `SmartPlay` call at line 376, I tried to change the reference to be `window.SmartPlay(...` and it appeared to work with the build file was minified (see example below).  
<img width="744" alt="Screen Shot 2020-08-28 at 10 09 44 AM" src="https://user-images.githubusercontent.com/31102355/91577434-ae3ce480-e916-11ea-9ca1-f735cf85dbe1.png">


Can you give this a try on your end?

Reply:
We use callback functions to give customers the flexibility to use callbacks should certain `events` occur. For example, customers can define the function sc_smartIntxtStart() on their webpage, which will be called by our player when the promotional video starts and the event gets fired. 

> The issue with the minified build is that it is likely seeing these variables never get assigned an actual value.

Do you mean this variables?
```
  // eslint-disable-next-line camelcase
  var sc_smartIntxtStart;
  // eslint-disable-next-line camelcase
  var sc_smartIntxtNoad;
  // eslint-disable-next-line camelcase
  var sc_smartIntxtEnd;
  var SmartPlay;
```
`If yes:` We added them back then on your hint due to `lint` errors, actually we don't need them and can leave them out. Or is it possible that exactly these variables confuse the minifier?
`If not:`  Do you mean the functions?

Here is the test site only with added 4 functions, I've tested it with `window.` before and it seems to work as well but I would prefer to let them be as they are and use this solution only if no other is available:
[https://gt-adtech.de/var/www/testsite/prod/prebid_test.html](https://gt-adtech.de/var/www/testsite/prod/prebid_test.html)
Could you please check if there are options for the minifier to ignore these 4 functions or could be the solution to remove the above not needed variables?

Thanks

Reply:
Hi @jsnellbaker,
did you already have time to have a look at it? We want to fix the problem as soon as possible, because the current version does not work like this.

Thanks

Reply:
Hi @Skylinar 

I meant callback functions altogether.  We can't have improperly defined functions within the code since it could lead to the chance that someone doesn't define the overridden function properly on their site and it would cause a hard undefined exception error in prebid - interfering with the rest of the auction.

If they're values that can be overwritten by the page so the publisher can customize what happens, then maybe this part of the logic could be handled inside the script itself instead of passing it in as part of a tentative config.  

By removing them from being in an 'undefined'/uninitialized state in the adapter, it would avoid the lint errors and confusing part of the logic (since on its own, it's not clear where these functions are coming from or what they're doing).

I understand if there may be other reasons why this may not be an option (or at least something quick to implement).  
In this case, I think you'll have to read these functions from the `window` to preserve their proper function status.  I don't know of anything else we can do in the minifying step to just skip over this part of the code.

Reply:
Hi @jsnellbaker,
I've created a new PR where we call it with `window`. Could you please check?

#5689

Thank you!